### PR TITLE
[WIP] Wide & Deep migration to PyTorch

### PR DESCRIPTION
**Pull request by daviddavo**
### Description

Migrating Wide & Deep out of tensorflow.

**Note**: Previously I have only used models from high level libraries like Keras. I'm doing this to learn PyTorch, so feel free to give me any pointers or even scrap everything if it is not useful.

### Related Issues

- #2072 

### References

- [The paper that defines the model](https://dl.acm.org/doi/10.1145/2988450.2988454)
- [A notebook from matanivanov which defines a PyTorch implementation](https://www.kaggle.com/code/matanivanov/wide-deep-learning-for-recsys-with-pytorch)
- [A full library of wide-deep models](https://github.com/jrzaurin/pytorch-widedeep)

### Checklist:

- [x] I have followed the [contribution guidelines](CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [x] I have [signed the commits](https://github.com/recommenders-team/recommenders/wiki/How-to-sign-commits), e.g. `git commit -s -m "your commit message"`. 
- [ ] This PR is being made to `staging branch` AND NOT TO `main branch`.

### CODE REVIEW

**Comment by miguelgfierro** [comment]
Sorry @daviddavo I pressed the wrong action. I changed the PR to ready for review because we are having some problems with the tests.

**Comment by daviddavo** [comment]
NP, I still have quite some work to do

**Comment by miguelgfierro** [comment]
FYI @daviddavo  the tests should be working now after https://github.com/recommenders-team/recommenders/pull/2169

**Comment by daviddavo** [comment]
The tensorflow estimators approach currently used by recommenders uses a binary regressor (default value of `n_classes`). To get the recommendations, all user-item pairs are used as input (created using the `user_item_pairs` method).

On the PyTorch notebook the output is not binary, instead there is a class for each movie. The aim of this notebook is to predict the next movie to watch, and not to make a top-k recommendations, a different problem.

The question is, **what does the original paper do?** Does it output a single scalar? Or does it output a vector with a value corresponding to each item? As I understand it, it should be a single value ( $P\left(Y=1\|X\right)=\sigma\left(...\right)$ ), but that notebook made me doubt.

Nevertheless, I think I will modify the current model so the "head" returns a scalar and to get the top-k recommendations we pass all the possible user-item pairs, as the recommenders' tensorflow implementation does.

Edit: [NVIDIA's deep learning examples also output a single value](https://github.com/NVIDIA/DeepLearningExamples/blob/729963dd47e7c8bd462ad10bfac7a7b0b604e6dd/TensorFlow2/Recommendation/WideAndDeep/trainer/model/widedeep.py#L104). My doubts have been resolved but I'll keep this post as some kind of documentation. I'll finish the model and do the training soon.

**Comment by daviddavo** [comment]
Loss function decreases over time in my jupyter notebook. the only thing remaining is the "software engineering" part

**Comment by daviddavo** [comment]
Now that I'm testing it with the full 100k dataset, I realized its very slow. I will profile it next weekend, but I have a hunch that the problem is the DataLoader, which uses a lot of slow `.loc`s

**Comment by miguelgfierro** [comment]
@@daviddavo how are things, I would like to ask whether you would be continuing with this work.

**Comment by miguelgfierro** [comment]
@daviddavo thanks David!
FYI @loomlike

**Comment by anargyri** [comment]
> Hi! Sorry for taking so long to reply! I changed jobs and moved to another city, so I did not have time for hobbies, sorry again for leaving you hanging.
> 
> I don't think I will have time in the near future, so it is fine if anyone else wants to keep working on this. If it remains open when I have time, I will retake it.

Thanks for the contributions @daviddavo ! Feel welcome to give feedback or contribute to the repo any time you have the time!

**Review by loomlike — APPROVED** [review]
